### PR TITLE
x86mmu: kernel: Validate existing APIs

### PR DESCRIPTION
**Pull request by akhileshupadhyay**
Testcase developed x86 specific to validate API for userbuffer.

Signed-off-by: Akhilesh Kumar Upadhyay <akhilesh.kumarx.upadhyay@intel.com>

### CODE REVIEW

File: tests/kernel/userbufer_validate/src/main.c [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+extern void test_multi_page_writeable(void);
+extern void test_multi_page_user_supervisor_rw(void);
+extern void test_multi_page_supervisor_rw(void);
+extern void test_multi_page_user_rw(void);
+extern void test_multi_page_supervisor_rw_user(void);
+extern void test_multi_page_supervisor_rw_supervisor(void);
+/*test case nain entry*/
+void test_main(void *p1, void *p2, void *p3)
+{
+		ztest_test_suite(userbuffer_validate,
```
Comment (tbursztyka) [inline tests/kernel/userbufer_validate/src/main.c]: 1 tab only for indentation on all these lines.

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+
+	return TC_PASS;
+}
+
+static int buffer_supervisor_rw(void)
+{
+	/* user/supervisor test */
+	x86_mmu_set_pte_us_bit((void *)0x00405000, 0); /* supervisor */
+	status = _x86_mmu_buffer_validate((void *)0x00405000,
+			10, 0x1 | 0x2); /*  should fail */
```
Comment (tbursztyka) [inline tests/kernel/userbufer_validate/src/userbuffer_validate.c]: parameters should be aligned to last '(' + 1 space, so here the 10, ... should be right belo (void *...

and the comment /* should fail */, put it on its own line.

Apply that rule everywhere

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+	}
+
+	return TC_PASS;
+}
+
+void test_buffer_rw_read(void)
+{
+	zassert_true(buffer_rw_read() == TC_PASS, NULL);
+}
+void test_buffer_writeable_write(void)
```
Comment (tbursztyka) [inline tests/kernel/userbufer_validate/src/userbuffer_validate.c]: always break a line before creating a new function. Apply that everywhere

**Review by tbursztyka — CHANGES_REQUESTED** [review]

**Review by pfalcon — CHANGES_REQUESTED** [review]
This patch is named as if it introduces "API to validate user buffer". But it does not, it just adds a test for an existing API. Please make sure that caption corresponds to the content.

File: tests/kernel/userbufer_validate/src/main.c (hunk 1 above)
Comment (akhileshupadhyay) [reply tests/kernel/userbufer_validate/src/main.c]: 1 tab indentation has been corrected.

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c (hunk 2 above)
Comment (akhileshupadhyay) [reply tests/kernel/userbufer_validate/src/userbuffer_validate.c]: breaking rule applied and corrected everywhere.

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c (hunk 3 above)
Comment (akhileshupadhyay) [reply tests/kernel/userbufer_validate/src/userbuffer_validate.c]: 1 line break applied for all functions.

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+ * Below conditions will be tested accordingly
+ *
+ */
+
+static int buffer_rw_read(void)
+{
+	/* read write testing */
+	x86_mmu_set_pte_rw_bit((void *)0x00405000, 0);
+	status = _x86_mmu_buffer_validate((void *)0x00405000,
+			10, BUFF_WRITEABLE); /* should fail */
```
Comment (tbursztyka) [inline tests/kernel/userbufer_validate/src/userbuffer_validate.c]: missed this line.

**Review by tbursztyka — DISMISSED** [review]

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c (hunk 4 above)
Comment (akhileshupadhyay) [reply tests/kernel/userbufer_validate/src/userbuffer_validate.c]: the missed line corrected.

**Comment by andyross** [comment]
Can we clean up the commit message?  This isn't x86 specific, it's actually PC-specific and assumes qemu's memory map.  The API being tested names itself something more like "x86 mmu" and not "user buffer".  Please explain exactly what functionality is being tested (it looks like it's simply setting PDE/PTE bits directly and then validating that they're still set?) in the commit message.

Also: wouldn't a more effective/portable/bitrot-resistant way to do this be to set the protection modes, and then try to access the pages in violation of that and check that the resulting exception is delivered correctly?

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+	}
+
+	return TC_PASS;
+}
+
+/* tc for multiple page buffers */
+/* read write testing */
+static int multi_page_buffer_writeable(void)
+{
+	x86_mmu_set_pte_rw_bit((void *)0x00405000, 0);
```
Comment (andrewboie) [inline tests/kernel/userbufer_validate/src/userbuffer_validate.c]: why are these addresses hard-coded?
you need to programmatically select memory addresses to test.
there are linker variables defined to help you with this. see include/linker/linker-defs.h

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]
You need to remove *all* the hard-coded memory addresses. This test shall not depend specifically on QEMU's memory map. Use defined linker variables to determine what memory ranges exist for various regions.

**Review by pfalcon — DISMISSED** [review]
Thanks for updating the commit subject. (Didn't look further, just removing my -1.)

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c (hunk 5 above)
Comment (akhileshupadhyay) [reply tests/kernel/userbufer_validate/src/userbuffer_validate.c]: hard-coded addresses have been removed, also the linker variables have been used

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+ */
+
+#include <zephyr.h>
+#include <misc/printk.h>
+#include <mmustructs.h>
+#include <arch/x86/arch.h>
+#include <linker/linker-defs.h>
+#include <ztest.h>
+
+void x86_mmu_set_pte_rw_bit(void *addr, bool value)
```
Comment (andrewboie) [inline tests/kernel/userbufer_validate/src/userbuffer_validate.c]: Why do we have these functions here?
We already have _x86_mmu_set_flags(). Use that instead.
You can make these functions call _x86_mmu_set_flags() if you want. But you should not be poking the page tables directly in this file.

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+	u32_t pte = MMU_PAGE_NUM(addr);
+
+	volatile struct x86_mmu_page_table *pte_addr =
+		(struct x86_mmu_page_table *)
+		(X86_MMU_PD->entry[pde].pt.page_table << 12);
+
+	pte_addr->entry[pte].rw = (u32_t)value;
+}
+
+void x86_mmu_set_pte_us_bit(void *addr, bool value)
```
Comment (andrewboie) [inline tests/kernel/userbufer_validate/src/userbuffer_validate.c]: Don't need this function use _x86_mmu_set_flags()

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c [hunk 8]
```diff
@@ -0,0 +1,401 @@
...
+	u32_t pde = MMU_PDE_NUM(addr);
+	u32_t pte = MMU_PAGE_NUM(addr);
+	volatile struct x86_mmu_page_table *pte_addr =
+		(struct x86_mmu_page_table *)
+		(X86_MMU_PD->entry[pde].pt.page_table << 12);
+
+	pte_addr->entry[pte].us = (u32_t)value;
+}
+
+void x86_mmu_set_pde_us_bit(void *addr, bool value)
```
Comment (andrewboie) [inline tests/kernel/userbufer_validate/src/userbuffer_validate.c]: Don't need this function use _x86_mmu_set_flags()

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c [hunk 9]
```diff
@@ -0,0 +1,401 @@
...
+	}
+
+	return TC_PASS;
+}
+
+static int buffer_writeable_write(void)
+{
+	x86_mmu_set_pte_rw_bit((void *)__kernel_ram_start, 1);
+	status = _x86_mmu_buffer_validate((void *)__kernel_ram_start,
+					   10, 0x1);
```
Comment (andrewboie) [inline tests/kernel/userbufer_validate/src/userbuffer_validate.c]: Please do not use hard-coded constants when we have macros. Replace 0x1 with BUF_WRITEABLE here and elsewhere

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]

**Comment by akhileshupadhyay** [comment]
removed x86_mmu_set_pte_us_bit,  x86_mmu_set_pte_us_bit, x86_mmu_set_pde_us_bit  functions.
Used  _x86_mmu_set_flags() instead. Also the macro is used for masks and flags.

**Review by AdithyaBaglody — CHANGES_REQUESTED** [review]
Test cases for Validating  multiple pages  are missing. Each of the  pages should be configured with different page table options(i.e RW/US).

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]
This coverage is insufficient.

Right now, _x86_mmu_buffer_validate() is broken, it actually always returns true (nonzero) no matter what.

You are inconsistently checking if the return value is 0 or -EPERM. Please see PR #1441

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]
I'm making some changes to this API so hold off on this PR until those land.

**Comment by andrewboie** [comment]
hi, any update on this?
the APIs are now fixed.

**Comment by akhileshupadhyay** [comment]
am working on it, update after fixing errors.

**Comment by akhileshupadhyay** [comment]
@andrewboie. I have made the changes, included test coverage.

**Comment by akhileshupadhyay** [comment]
@tbursztyka. The proper indentation in missed lines are restructured.

**Comment by andrewboie** [comment]
@AdithyaBaglody can you review this?

File: tests/kernel/userbufer_validate/src/main.c [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+extern void test_buffer_rw_read(void);
+extern void test_buffer_writeable_write(void);
+extern void test_buffer_readable_read(void);
+extern void test_buffer_readable_write(void);
+extern void test_buffer_supervisor_rw(void);
+extern void test_buffer_supervisor_w(void);
+extern void test_buffer_user_rw_user(void);
+extern void test_buffer_user_rw_supervisor(void);
+/*test case nain entry*/
+void test_main(void *p1, void *p2, void *p3)
```
Comment (andrewboie) [inline tests/kernel/userbufer_validate/src/main.c]: why is this in a separate file???

Comment (akhileshupadhyay) [reply tests/kernel/userbufer_validate/src/main.c]: This will make the code more legible and easy. Also some of the other codes follow same.

**Comment by akhileshupadhyay** [comment]
@AdithyaBaglody . Changes related to multi pages. Removed all the hard coded values. Lines are aligned with one line gap after every new function starts.

**Comment by akhileshupadhyay** [comment]
@AdithyaBaglody  I have added more test cases. This will cover multi pages test.

**Review by AdithyaBaglody — APPROVED** [review]

**Comment by akhileshupadhyay** [comment]
@andrewboie could you please review the changes requested by you.

**Review by kumarvikash1 — APPROVED** [review]
Changes looks fine now, i can see "multi page buffer", "code coverage" and line breaking has been addressed.

**Comment by kumarvikash1** [comment]
@andrewboie  @tbursztyka   : Can you please review the changes. This PR is planned for R-1.10.

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c [hunk 11]
```diff
@@ -0,0 +1,524 @@
...
+ * else return 0.
+ * Below conditions will be tested accordingly
+ *
+ */
+
+/* read write testing */
+static int buffer_rw_read(void)
+{
+	PRESET_PAGE_1_VALUE;
+	_x86_mmu_set_flags(ADDR_PAGE_1,
```
Comment (tbursztyka) [inline tests/kernel/userbufer_validate/src/userbuffer_validate.c]: break a line after PRESET_... You do it sometime and sometime not. Apply that eveywhere

**Review by tbursztyka — COMMENTED** [review]
tiny style issue. rest is fine to me.

File: tests/kernel/userbufer_validate/src/userbuffer_validate.c (hunk 11 above)
Comment (akhileshupadhyay) [reply tests/kernel/userbufer_validate/src/userbuffer_validate.c]: @tbursztyka . Could you please check the style issue is addressed and fixed

**Review by tyrantlizard77 — APPROVED** [review]

**Pull request merged**